{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d01b181d",
   "metadata": {},
   "source": [
    "# let's have a look at vgsc snps\n",
    "- look at gamb_colu and arab separately"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3f5bf2",
   "metadata": {},
   "source": [
    "### setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "72999054",
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import bokeh.plotting as bkplt\n",
    "import bokeh.models as bkmod\n",
    "import bokeh.layouts as bklay\n",
    "from bokeh.layouts import column\n",
    "import bokeh.io as bkio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3a44d173",
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(\"simplecache::gs://vo_agam_release/\", \n",
    "                          simplecache=dict(cache_storage=\"gcs_cache\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "74544e04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#the transcript we have used previously for vgsc\n",
    "vgsc = \"AGAP004707-RD\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8475ce21",
   "metadata": {},
   "source": [
    "\n",
    "### gamb_colu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5830a4bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get snp effects for gamb_colu\n",
    "veff_gc_df = ag3.snp_effects(transcript=vgsc, site_mask='gamb_colu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7946042b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'partner_sample_id', 'contributor', 'country', 'location',\n",
       "       'year', 'month', 'latitude', 'longitude', 'sex_call', 'sample_set',\n",
       "       'release', 'aim_fraction_colu', 'aim_fraction_arab',\n",
       "       'species_gambcolu_arabiensis', 'species_gambiae_coluzzii', 'species'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# meta data\n",
    "metadata_df = ag3.sample_metadata(sample_sets=\"v3_wild\")\n",
    "metadata_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3b0168b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Angola', 'Burkina Faso', 'Democratic Republic of Congo',\n",
       "       'Central African Republic', \"Cote d'Ivoire\", 'Cameroon', 'Mayotte',\n",
       "       'Gabon', 'Ghana', 'Gambia, The', 'Guinea', 'Mali',\n",
       "       'Equatorial Guinea', 'Guinea-Bissau', 'Kenya', 'Malawi',\n",
       "       'Mozambique', 'Tanzania', 'Uganda'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define populations\n",
    "metadata_df.country.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1c7d17ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25,\n",
       "                          country                        species  year\n",
       " 0                         Angola                       coluzzii    81\n",
       " 2                   Burkina Faso                       coluzzii   135\n",
       " 3                   Burkina Faso                        gambiae   157\n",
       " 6                       Cameroon                       coluzzii    26\n",
       " 7                       Cameroon                        gambiae   416\n",
       " 9       Central African Republic                        gambiae    55\n",
       " 10                 Cote d'Ivoire                       coluzzii    80\n",
       " 11  Democratic Republic of Congo                        gambiae    76\n",
       " 13                         Gabon                        gambiae    69\n",
       " 14                   Gambia, The                       coluzzii   169\n",
       " 15                   Gambia, The                        gambiae    69\n",
       " 16                   Gambia, The  intermediate_gambiae_coluzzii    41\n",
       " 17                         Ghana                       coluzzii    64\n",
       " 18                         Ghana                        gambiae    36\n",
       " 20                        Guinea                        gambiae   123\n",
       " 22                 Guinea-Bissau                        gambiae    29\n",
       " 23                 Guinea-Bissau  intermediate_gambiae_coluzzii    72\n",
       " 25                         Kenya                        gambiae    28\n",
       " 26                         Kenya  intermediate_gambiae_coluzzii    45\n",
       " 29                          Mali                       coluzzii    91\n",
       " 30                          Mali                        gambiae   131\n",
       " 32                       Mayotte                        gambiae    23\n",
       " 33                    Mozambique                        gambiae    74\n",
       " 35                      Tanzania                        gambiae    68\n",
       " 38                        Uganda                        gambiae   207)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look at unique country+species counts - this will be our initial resolution\n",
    "# drop arabiensis and pops smaller than 20\n",
    "uni_df = metadata_df[['country', 'species', 'year']].groupby(['country','species']).count()\n",
    "uni_df.reset_index(inplace=True)\n",
    "uni_df = uni_df[uni_df['species'].map(lambda x: str(x)!=\"arabiensis\")]\n",
    "uni_df = uni_df[uni_df['year'].map(lambda x: x >= 20)]\n",
    "len(uni_df), uni_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9bfaa916",
   "metadata": {},
   "outputs": [],
   "source": [
    "#build dictionary\n",
    "populations_gc = {c+\"_\"+s: \"country == '''\"+c+\"''' and species == '\"+s+\"'\" for (c,s) in zip(list(uni_df.country), list(uni_df.species))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7bf415df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Angola_coluzzii': \"country == '''Angola''' and species == 'coluzzii'\",\n",
       " 'Burkina Faso_coluzzii': \"country == '''Burkina Faso''' and species == 'coluzzii'\",\n",
       " 'Burkina Faso_gambiae': \"country == '''Burkina Faso''' and species == 'gambiae'\",\n",
       " 'Cameroon_coluzzii': \"country == '''Cameroon''' and species == 'coluzzii'\",\n",
       " 'Cameroon_gambiae': \"country == '''Cameroon''' and species == 'gambiae'\",\n",
       " 'Central African Republic_gambiae': \"country == '''Central African Republic''' and species == 'gambiae'\",\n",
       " \"Cote d'Ivoire_coluzzii\": \"country == '''Cote d'Ivoire''' and species == 'coluzzii'\",\n",
       " 'Democratic Republic of Congo_gambiae': \"country == '''Democratic Republic of Congo''' and species == 'gambiae'\",\n",
       " 'Gabon_gambiae': \"country == '''Gabon''' and species == 'gambiae'\",\n",
       " 'Gambia, The_coluzzii': \"country == '''Gambia, The''' and species == 'coluzzii'\",\n",
       " 'Gambia, The_gambiae': \"country == '''Gambia, The''' and species == 'gambiae'\",\n",
       " 'Gambia, The_intermediate_gambiae_coluzzii': \"country == '''Gambia, The''' and species == 'intermediate_gambiae_coluzzii'\",\n",
       " 'Ghana_coluzzii': \"country == '''Ghana''' and species == 'coluzzii'\",\n",
       " 'Ghana_gambiae': \"country == '''Ghana''' and species == 'gambiae'\",\n",
       " 'Guinea_gambiae': \"country == '''Guinea''' and species == 'gambiae'\",\n",
       " 'Guinea-Bissau_gambiae': \"country == '''Guinea-Bissau''' and species == 'gambiae'\",\n",
       " 'Guinea-Bissau_intermediate_gambiae_coluzzii': \"country == '''Guinea-Bissau''' and species == 'intermediate_gambiae_coluzzii'\",\n",
       " 'Kenya_gambiae': \"country == '''Kenya''' and species == 'gambiae'\",\n",
       " 'Kenya_intermediate_gambiae_coluzzii': \"country == '''Kenya''' and species == 'intermediate_gambiae_coluzzii'\",\n",
       " 'Mali_coluzzii': \"country == '''Mali''' and species == 'coluzzii'\",\n",
       " 'Mali_gambiae': \"country == '''Mali''' and species == 'gambiae'\",\n",
       " 'Mayotte_gambiae': \"country == '''Mayotte''' and species == 'gambiae'\",\n",
       " 'Mozambique_gambiae': \"country == '''Mozambique''' and species == 'gambiae'\",\n",
       " 'Tanzania_gambiae': \"country == '''Tanzania''' and species == 'gambiae'\",\n",
       " 'Uganda_gambiae': \"country == '''Uganda''' and species == 'gambiae'\"}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "populations_gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e7a3faa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get allele freqs\n",
    "allele_freq_df = ag3.snp_allele_frequencies(transcript=vgsc, \n",
    "                           populations=populations_gc, \n",
    "                           site_mask=\"gamb_colu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d3fcbc68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge our two dfs\n",
    "merged_df = pd.merge(veff_gc_df, allele_freq_df, on=['position', 'ref_allele', 'alt_allele'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2c3ca459",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>...</th>\n",
       "      <th>Guinea-Bissau_intermediate_gambiae_coluzzii</th>\n",
       "      <th>Kenya_gambiae</th>\n",
       "      <th>Kenya_intermediate_gambiae_coluzzii</th>\n",
       "      <th>Mali_coluzzii</th>\n",
       "      <th>Mali_gambiae</th>\n",
       "      <th>Mayotte_gambiae</th>\n",
       "      <th>Mozambique_gambiae</th>\n",
       "      <th>Tanzania_gambiae</th>\n",
       "      <th>Uganda_gambiae</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2358161</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Acc</td>\n",
       "      <td>Gcc</td>\n",
       "      <td>2.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2358163</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SYNONYMOUS_CODING</td>\n",
       "      <td>LOW</td>\n",
       "      <td>acC</td>\n",
       "      <td>acG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2358164</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Gaa</td>\n",
       "      <td>Aaa</td>\n",
       "      <td>3.0</td>\n",
       "      <td>E</td>\n",
       "      <td>K</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2358168</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>gAc</td>\n",
       "      <td>gTc</td>\n",
       "      <td>4.0</td>\n",
       "      <td>D</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2358168</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>gAc</td>\n",
       "      <td>gGc</td>\n",
       "      <td>4.0</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14117</th>\n",
       "      <td>2431601</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>cGa</td>\n",
       "      <td>cAa</td>\n",
       "      <td>2114.0</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14118</th>\n",
       "      <td>2431601</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>cGa</td>\n",
       "      <td>cTa</td>\n",
       "      <td>2114.0</td>\n",
       "      <td>R</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002415</td>\n",
       "      <td>0.002415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14119</th>\n",
       "      <td>2431609</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Gat</td>\n",
       "      <td>Tat</td>\n",
       "      <td>2117.0</td>\n",
       "      <td>D</td>\n",
       "      <td>Y</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14120</th>\n",
       "      <td>2431613</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>gTc</td>\n",
       "      <td>gCc</td>\n",
       "      <td>2118.0</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14121</th>\n",
       "      <td>2431614</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SYNONYMOUS_CODING</td>\n",
       "      <td>LOW</td>\n",
       "      <td>gtC</td>\n",
       "      <td>gtT</td>\n",
       "      <td>2118.0</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006944</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006944</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14122 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       position ref_allele alt_allele                 effect    impact  \\\n",
       "0       2358161          A          G  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "1       2358163          C          G      SYNONYMOUS_CODING       LOW   \n",
       "2       2358164          G          A  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "3       2358168          A          T  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "4       2358168          A          G  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "...         ...        ...        ...                    ...       ...   \n",
       "14117   2431601          G          A  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "14118   2431601          G          T  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "14119   2431609          G          T  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "14120   2431613          T          C  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "14121   2431614          C          T      SYNONYMOUS_CODING       LOW   \n",
       "\n",
       "      ref_codon alt_codon  aa_pos ref_aa alt_aa  ...  \\\n",
       "0           Acc       Gcc     2.0      T      A  ...   \n",
       "1           acC       acG     2.0      T      T  ...   \n",
       "2           Gaa       Aaa     3.0      E      K  ...   \n",
       "3           gAc       gTc     4.0      D      V  ...   \n",
       "4           gAc       gGc     4.0      D      G  ...   \n",
       "...         ...       ...     ...    ...    ...  ...   \n",
       "14117       cGa       cAa  2114.0      R      Q  ...   \n",
       "14118       cGa       cTa  2114.0      R      L  ...   \n",
       "14119       Gat       Tat  2117.0      D      Y  ...   \n",
       "14120       gTc       gCc  2118.0      V      A  ...   \n",
       "14121       gtC       gtT  2118.0      V      V  ...   \n",
       "\n",
       "      Guinea-Bissau_intermediate_gambiae_coluzzii  Kenya_gambiae  \\\n",
       "0                                        0.000000            0.0   \n",
       "1                                        0.000000            0.0   \n",
       "2                                        0.000000            0.0   \n",
       "3                                        0.000000            0.0   \n",
       "4                                        0.000000            0.0   \n",
       "...                                           ...            ...   \n",
       "14117                                    0.000000            0.0   \n",
       "14118                                    0.000000            0.0   \n",
       "14119                                    0.000000            0.0   \n",
       "14120                                    0.000000            0.0   \n",
       "14121                                    0.006944            0.0   \n",
       "\n",
       "       Kenya_intermediate_gambiae_coluzzii  Mali_coluzzii  Mali_gambiae  \\\n",
       "0                                      0.0            0.0           0.0   \n",
       "1                                      0.0            0.0           0.0   \n",
       "2                                      0.0            0.0           0.0   \n",
       "3                                      0.0            0.0           0.0   \n",
       "4                                      0.0            0.0           0.0   \n",
       "...                                    ...            ...           ...   \n",
       "14117                                  0.0            0.0           0.0   \n",
       "14118                                  0.0            0.0           0.0   \n",
       "14119                                  0.0            0.0           0.0   \n",
       "14120                                  0.0            0.0           0.0   \n",
       "14121                                  0.0            0.0           0.0   \n",
       "\n",
       "       Mayotte_gambiae  Mozambique_gambiae  Tanzania_gambiae  Uganda_gambiae  \\\n",
       "0                  0.0            0.000000               0.0        0.000000   \n",
       "1                  0.0            0.000000               0.0        0.000000   \n",
       "2                  0.0            0.000000               0.0        0.000000   \n",
       "3                  0.0            0.000000               0.0        0.000000   \n",
       "4                  0.0            0.000000               0.0        0.000000   \n",
       "...                ...                 ...               ...             ...   \n",
       "14117              0.0            0.000000               0.0        0.000000   \n",
       "14118              0.0            0.000000               0.0        0.002415   \n",
       "14119              0.0            0.027027               0.0        0.000000   \n",
       "14120              0.0            0.000000               0.0        0.000000   \n",
       "14121              0.0            0.000000               0.0        0.000000   \n",
       "\n",
       "        maximum  \n",
       "0      0.004065  \n",
       "1      0.019231  \n",
       "2      0.006579  \n",
       "3      0.001202  \n",
       "4      0.001202  \n",
       "...         ...  \n",
       "14117  0.017241  \n",
       "14118  0.002415  \n",
       "14119  0.027027  \n",
       "14120  0.006173  \n",
       "14121  0.006944  \n",
       "\n",
       "[14122 rows x 37 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "56ae23b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter\n",
    "filter_df  = merged_df[(merged_df.impact == \"HIGH\") | (merged_df.impact == \"MODERATE\" )].copy()\n",
    "freq_filtered_df = filter_df[filter_df.maximum > 0].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "49e798c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting stuff\n",
    "freq_filtered_df['pos_Mbp'] = freq_filtered_df.position/1e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "76e66e7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneset = ag3.geneset()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3342ec5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>49364325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>157348</td>\n",
       "      <td>186936</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>157348</td>\n",
       "      <td>181305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677-RA</td>\n",
       "      <td>AGAP004677</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>157348</td>\n",
       "      <td>157495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>157348</td>\n",
       "      <td>157623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677-RA</td>\n",
       "      <td>AGAP004677-RB-E4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196140</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>47932</td>\n",
       "      <td>48111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196141</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>47932</td>\n",
       "      <td>48138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>AGAP029375-RA-E2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196142</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>48112</td>\n",
       "      <td>48138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGAP029375-PA</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196143</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>48301</td>\n",
       "      <td>48385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>AGAP029375-RA-E3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196144</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>48301</td>\n",
       "      <td>48385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGAP029375-PA</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196145 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid      source             type   start      stop  score  \\\n",
       "0               2L  VectorBase       chromosome       1  49364325    NaN   \n",
       "1               2L  VectorBase             gene  157348    186936    NaN   \n",
       "2               2L  VectorBase             mRNA  157348    181305    NaN   \n",
       "3               2L  VectorBase  three_prime_UTR  157348    157495    NaN   \n",
       "4               2L  VectorBase             exon  157348    157623    NaN   \n",
       "...            ...         ...              ...     ...       ...    ...   \n",
       "196140  Y_unplaced  VectorBase   five_prime_UTR   47932     48111    NaN   \n",
       "196141  Y_unplaced  VectorBase             exon   47932     48138    NaN   \n",
       "196142  Y_unplaced  VectorBase              CDS   48112     48138    NaN   \n",
       "196143  Y_unplaced  VectorBase             exon   48301     48385    NaN   \n",
       "196144  Y_unplaced  VectorBase              CDS   48301     48385    NaN   \n",
       "\n",
       "       strand  phase     feature_id      parent_id              Name  \n",
       "0         NaN    NaN             2L            NaN               NaN  \n",
       "1           -    NaN     AGAP004677            NaN               NaN  \n",
       "2           -    NaN  AGAP004677-RA     AGAP004677               NaN  \n",
       "3           -    NaN            NaN  AGAP004677-RA               NaN  \n",
       "4           -    NaN            NaN  AGAP004677-RA  AGAP004677-RB-E4  \n",
       "...       ...    ...            ...            ...               ...  \n",
       "196140      +    NaN            NaN  AGAP029375-RA               NaN  \n",
       "196141      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E2  \n",
       "196142      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN  \n",
       "196143      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E3  \n",
       "196144      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN  \n",
       "\n",
       "[196145 rows x 11 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "05444117",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_transcript_variation(freq_df, transcript, populations, width=800, height=400):\n",
    "\n",
    "    # SCATTER\n",
    "    # tooltips\n",
    "    tooltips1 = [\n",
    "        (\"population\", '@population'),\n",
    "        (\"frequency\", '@frequency'),\n",
    "        (\"position\", '@position'),\n",
    "        (\"aa change\", '@aa_change'),\n",
    "    ]\n",
    "\n",
    "    fig1 = bkplt.figure(\n",
    "        title=f'Transcript - {transcript}',\n",
    "        tools='xpan,xzoom_in,xzoom_out,xwheel_zoom,reset,hover',\n",
    "        active_scroll='xwheel_zoom',\n",
    "        active_drag='xpan',\n",
    "        plot_width=width, \n",
    "        plot_height=height, \n",
    "        tooltips=tooltips1,\n",
    "        toolbar_location=\"above\")\n",
    "\n",
    "    for pop in populations:\n",
    "        df = freq_df.copy()\n",
    "        df['frequency'] = df[pop]\n",
    "        df['population'] = pop\n",
    "        fig1.circle(\"pos_Mbp\", \n",
    "                    pop, size=10,  \n",
    "                    alpha=0.8, \n",
    "                    source=df,\n",
    "                    legend_label=pop)\n",
    "\n",
    "    fig1.yaxis.axis_label = f'population nucleotide frequency'\n",
    "    fig1.xaxis.visible = False\n",
    "#     fig1.legend.location = \"top_left\"\n",
    "    fig1.add_layout(fig1.legend[0], 'right')\n",
    "    fig1.legend.click_policy=\"hide\"\n",
    "    \n",
    "    # GENE MODEL\n",
    "    df_geneset = ag3.geneset()\n",
    "    df_transcripts = df_geneset.query(f\"feature_id == '{transcript}'\")\n",
    "\n",
    "    # define tooltips for hover\n",
    "    tooltips2 = [\n",
    "        (\"Type\", '@type'),\n",
    "        (\"Location\", '@seqid:@start{,}..@stop{,}'),\n",
    "    ]\n",
    "\n",
    "    # make a figure\n",
    "    fig2 = bkplt.figure(   \n",
    "        plot_width=width, \n",
    "        plot_height=int(height / 4),\n",
    "        tools='xpan,xzoom_in,xzoom_out,xwheel_zoom,reset,hover',\n",
    "        #toolbar_location='above',\n",
    "        active_scroll='xwheel_zoom',\n",
    "        active_drag='xpan',\n",
    "        tooltips=tooltips2,\n",
    "        x_range=fig1.x_range,\n",
    "    )\n",
    "\n",
    "    # plot the transcripts\n",
    "\n",
    "\n",
    "    data = df_geneset.query(f\"parent_id == '{transcript}'\").copy()\n",
    "    data['left'] = data['start'] / 1e6  # plot in Mbp coordinates\n",
    "    data['right'] = data['stop'] / 1e6  # plot in Mbp coordinates\n",
    "    data['bottom'] = 0 - 0.4\n",
    "    data['top'] = 0 + 0.4\n",
    "\n",
    "    # plot exons\n",
    "    exons = data.query(\"type == 'exon'\")\n",
    "    fig2.quad(bottom='bottom', \n",
    "              top='top', \n",
    "              left='left', \n",
    "              right='right',\n",
    "              source=exons, \n",
    "              fill_color=None, \n",
    "              line_color='black', \n",
    "              line_width=.5, \n",
    "              fill_alpha=0)\n",
    "\n",
    "    # plot introns\n",
    "    for l, r in zip(exons[:-1]['right'], exons[1:]['left']):\n",
    "        m = (l + r) / 2\n",
    "        fig2.line([l, m, r], [0, 0 +.1, 0], line_width=1, line_color=\"black\")\n",
    "\n",
    "    # plot UTRs\n",
    "    fig2.quad(bottom='bottom', top='top', left='left', right='right',\n",
    "             source=data.query(\"type == 'five_prime_UTR'\"), \n",
    "             fill_color='green', line_width=0, fill_alpha=.5)\n",
    "    fig2.quad(bottom='bottom', top='top', left='left', right='right',\n",
    "             source=data.query(\"type == 'three_prime_UTR'\"), \n",
    "             fill_color='red', line_width=0, fill_alpha=.5)\n",
    "\n",
    "    # plot CDSs\n",
    "    fig2.quad(bottom='bottom', top='top', left='left', right='right',\n",
    "             source=data.query(\"type == 'CDS'\"), \n",
    "             fill_color='blue', line_width=0, fill_alpha=.5)\n",
    "\n",
    "    fig2.xaxis.axis_label = f'Position (Mbp)'\n",
    "    fig2.y_range = bkmod.Range1d(-.5, len(df_transcripts)-.5)\n",
    "    fig2.xaxis.axis_label = f'Contig {df_transcripts.seqid.item()} position (Mbp)'\n",
    "#     yticks = list(range(len(df_transcripts)))\n",
    "#     yticklabels = df_transcripts.index\n",
    "#     fig2.yaxis.ticker = yticks\n",
    "#     fig2.yaxis.major_label_overrides = {k: v for k, v in zip(yticks, yticklabels)}\n",
    "    fig2.toolbar.logo = None \n",
    "    fig2.toolbar_location = None\n",
    "\n",
    "    bkplt.show(column(fig1, fig2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5d62c783",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_transcript_variation(freq_df=freq_filtered_df, transcript=vgsc, populations=populations_gc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80a622dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# todo - fix ytick brokeness and add plot into notebook\n",
    "# - we will need colours for each pop...JOY"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
