{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "045fc990",
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "# import allel\n",
    "# import numpy as np\n",
    "# from tqdm.auto import tqdm\n",
    "# from tqdm.dask import TqdmCallback\n",
    "# import numpy as np\n",
    "# import bokeh.io as bkio\n",
    "# bkio.reset_output()\n",
    "# bkio.output_notebook(hide_banner=True)\n",
    "# import bokeh.plotting as bkplt\n",
    "# import bokeh.models as bkmod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "32b3615a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ag3 = malariagen_data.Ag3(results_cache=\"Git/results_cache\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "98f5a997",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <table class=\"malariagen-ag3\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\" colspan=\"2\">MalariaGEN Ag3 API client</th>\n",
       "                    </tr>\n",
       "                    <tr><td colspan=\"2\" style=\"text-align: left\">\n",
       "                        Please note that data are subject to terms of use,\n",
       "                        for more information see <a href=\"https://www.malariagen.net/data\">\n",
       "                        the MalariaGEN website</a> or contact data@malariagen.net.\n",
       "                        See also the <a href=\"https://malariagen.github.io/vector-data/ag3/api.html\">Ag3 API docs</a>.\n",
       "                    </td></tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Storage URL\n",
       "                        </th>\n",
       "                        <td>gs://vo_agam_release/</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Data releases available\n",
       "                        </th>\n",
       "                        <td>3.0</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Results cache\n",
       "                        </th>\n",
       "                        <td>/home/kai/Git/malariagen-data-python/Git/results_cache</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Cohorts analysis\n",
       "                        </th>\n",
       "                        <td>20220608</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Species analysis\n",
       "                        </th>\n",
       "                        <td>aim_20220528</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Site filters analysis\n",
       "                        </th>\n",
       "                        <td>dt_20200416</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Software version\n",
       "                        </th>\n",
       "                        <td>malariagen_data 6.0.1</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Client location\n",
       "                        </th>\n",
       "                        <td>England, GB</td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        "
      ],
      "text/plain": [
       "<MalariaGEN Ag3 API client>\n",
       "Storage URL             : gs://vo_agam_release/\n",
       "Data releases available : 3.0\n",
       "Results cache           : /home/kai/Git/malariagen-data-python/Git/results_cache\n",
       "Cohorts analysis        : 20220608\n",
       "Species analysis        : aim_20220528\n",
       "Site filters analysis   : dt_20200416\n",
       "Software version        : malariagen_data 6.0.1\n",
       "Client location         : England, GB\n",
       "---\n",
       "Please note that data are subject to terms of use,\n",
       "for more information see https://www.malariagen.net/data\n",
       "or contact data@malariagen.net. For API documentation see \n",
       "https://malariagen.github.io/vector-data/ag3/api.html"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "70176572",
   "metadata": {},
   "outputs": [],
   "source": [
    "release = \"3.0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cd068db1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_set</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>release</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG1000G-AO</td>\n",
       "      <td>81</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG1000G-BF-A</td>\n",
       "      <td>181</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG1000G-BF-B</td>\n",
       "      <td>102</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG1000G-BF-C</td>\n",
       "      <td>13</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG1000G-CD</td>\n",
       "      <td>76</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG1000G-CF</td>\n",
       "      <td>73</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG1000G-CI</td>\n",
       "      <td>80</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG1000G-CM-A</td>\n",
       "      <td>303</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG1000G-CM-B</td>\n",
       "      <td>97</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG1000G-CM-C</td>\n",
       "      <td>44</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AG1000G-FR</td>\n",
       "      <td>23</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AG1000G-GA-A</td>\n",
       "      <td>69</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AG1000G-GH</td>\n",
       "      <td>100</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AG1000G-GM-A</td>\n",
       "      <td>74</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AG1000G-GM-B</td>\n",
       "      <td>31</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AG1000G-GM-C</td>\n",
       "      <td>174</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AG1000G-GN-A</td>\n",
       "      <td>45</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AG1000G-GN-B</td>\n",
       "      <td>185</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AG1000G-GQ</td>\n",
       "      <td>10</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AG1000G-GW</td>\n",
       "      <td>101</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AG1000G-KE</td>\n",
       "      <td>86</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AG1000G-ML-A</td>\n",
       "      <td>60</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AG1000G-ML-B</td>\n",
       "      <td>71</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AG1000G-MW</td>\n",
       "      <td>41</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AG1000G-MZ</td>\n",
       "      <td>74</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AG1000G-TZ</td>\n",
       "      <td>300</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>290</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AG1000G-X</td>\n",
       "      <td>297</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_set  sample_count release\n",
       "0     AG1000G-AO            81     3.0\n",
       "1   AG1000G-BF-A           181     3.0\n",
       "2   AG1000G-BF-B           102     3.0\n",
       "3   AG1000G-BF-C            13     3.0\n",
       "4     AG1000G-CD            76     3.0\n",
       "5     AG1000G-CF            73     3.0\n",
       "6     AG1000G-CI            80     3.0\n",
       "7   AG1000G-CM-A           303     3.0\n",
       "8   AG1000G-CM-B            97     3.0\n",
       "9   AG1000G-CM-C            44     3.0\n",
       "10    AG1000G-FR            23     3.0\n",
       "11  AG1000G-GA-A            69     3.0\n",
       "12    AG1000G-GH           100     3.0\n",
       "13  AG1000G-GM-A            74     3.0\n",
       "14  AG1000G-GM-B            31     3.0\n",
       "15  AG1000G-GM-C           174     3.0\n",
       "16  AG1000G-GN-A            45     3.0\n",
       "17  AG1000G-GN-B           185     3.0\n",
       "18    AG1000G-GQ            10     3.0\n",
       "19    AG1000G-GW           101     3.0\n",
       "20    AG1000G-KE            86     3.0\n",
       "21  AG1000G-ML-A            60     3.0\n",
       "22  AG1000G-ML-B            71     3.0\n",
       "23    AG1000G-MW            41     3.0\n",
       "24    AG1000G-MZ            74     3.0\n",
       "25    AG1000G-TZ           300     3.0\n",
       "26    AG1000G-UG           290     3.0\n",
       "27     AG1000G-X           297     3.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag3.sample_sets(release=release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a12b7385",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_ghana = \"country == 'Ghana'\"\n",
    "contig=\"3L\" \n",
    "analysis=\"gamb_colu\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe5e993",
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3.count_samples(sample_query=query_ghana)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a71de886",
   "metadata": {},
   "outputs": [],
   "source": [
    "def h12_calibration(contig, \n",
    "                    analysis, \n",
    "                    sample_query,\n",
    "                    sample_sets, \n",
    "                    cohort_size=20,\n",
    "                    window_sizes=(100, 200, 500, 1000, 2000, 5000, 10000, 20000),\n",
    "                    random_seed=42):\n",
    "    \n",
    "    # access haplotypes\n",
    "    ds_haps = ag3.haplotypes(\n",
    "        region=contig, \n",
    "        sample_sets=sample_sets, \n",
    "        sample_query=sample_query, \n",
    "        analysis=analysis,\n",
    "        cohort_size=cohort_size,\n",
    "        random_seed=random_seed\n",
    "        \n",
    "    )\n",
    "    \n",
    "    gt = allel.GenotypeDaskArray(ds_haps['call_genotype'].data)\n",
    "    \n",
    "    # TODO - cache haplotype data \n",
    "    \n",
    "    ht = gt.to_haplotypes().compute()\n",
    "    \n",
    "    calibration_runs = list()\n",
    "    for window_size in tqdm(window_sizes, desc=\"Compute H12\"):\n",
    "        h1, h12, h123, h2_h1 = allel.moving_garud_h(ht, size=window_size)\n",
    "        calibration_runs.append(h12)\n",
    "        \n",
    "    # TODO - cache calibration runs\n",
    "    \n",
    "    return calibration_runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cb30c876",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.86 ms, sys: 0 ns, total: 2.86 ms\n",
      "Wall time: 2.11 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "calibration_runs = ag3.h12_calibration(contig=contig, analysis=analysis, sample_query=query_ghana, \n",
    "                    sample_sets=release, cohort_size=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1488f4d4",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'100 is not a file in the archive'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_41103/673815678.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcalibration_runs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.cache/pypoetry/virtualenvs/malariagen-data-kqwLHFH8-py3.8/lib/python3.8/site-packages/numpy/lib/npyio.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    258\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mzip\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    259\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 260\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s is not a file in the archive\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    261\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: '100 is not a file in the archive'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68a2749b",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(calibration_runs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "070ad4ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bokeh.plotting as bkplt\n",
    "import bokeh.models as bkmod\n",
    "\n",
    "fig = bkplt.figure(plot_width=700, plot_height=400, x_axis_type=\"log\")\n",
    "\n",
    "fig.patch(x+x[::-1],q75+q25[::-1], alpha=0.75, line_width=2, legend_label=\"25-75%\")\n",
    "fig.patch(x+x[::-1],q95+q05[::-1], alpha=0.5, line_width=2, legend_label=\"5-95%\")\n",
    "fig.line(x,y, line_color='black', line_width=4, legend_label=\"median\")\n",
    "fig.circle(x, y, color='black',fill_color='black', size=8)\n",
    "\n",
    "\n",
    "fig.xaxis.ticker = x\n",
    "fig.x_range = bkmod.Range1d(100, 10000)\n",
    "fig.title = \"test\"\n",
    "bkplt.show(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "830ed422",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_h12_calibration(contig, \n",
    "                         analysis, \n",
    "                         sample_query,\n",
    "                         sample_sets,\n",
    "                         window_sizes=(100, 200, 500, 1000, 2000, 5000, 10000, 20000), \n",
    "                         cohort_size=20, \n",
    "                         random_seed=42, \n",
    "                         title=None):\n",
    "    \n",
    "    # get H12 values\n",
    "    calibration_runs = h12_calibration(\n",
    "        contig=contig, \n",
    "        analysis=analysis, \n",
    "        sample_query=sample_query, \n",
    "        sample_sets=sample_sets, \n",
    "        window_sizes=window_sizes,\n",
    "        cohort_size=cohort_size, \n",
    "        random_seed=random_seed\n",
    "    )\n",
    "\n",
    "    # compute summaries\n",
    "    q50 = [np.median(h12) for h12 in calibration_runs]\n",
    "    q25 = [np.percentile(h12, 25) for h12 in calibration_runs]\n",
    "    q75 = [np.percentile(h12, 75) for h12 in calibration_runs]\n",
    "    q05 = [np.percentile(h12, 5) for h12 in calibration_runs]\n",
    "    q95 = [np.percentile(h12, 95) for h12 in calibration_runs]\n",
    "\n",
    "    # make plot \n",
    "    fig = bkplt.figure(plot_width=700, plot_height=400, x_axis_type=\"log\")\n",
    "    fig.patch(window_sizes+window_sizes[::-1],q75+q25[::-1], alpha=0.75, line_width=2, legend_label=\"25-75%\")\n",
    "    fig.patch(window_sizes+window_sizes[::-1],q95+q05[::-1], alpha=0.5, line_width=2, legend_label=\"5-95%\")\n",
    "    fig.line(window_sizes, q50, line_color='black', line_width=4, legend_label=\"median\")\n",
    "    fig.circle(window_sizes, q50, color='black',fill_color='black', size=8)\n",
    "\n",
    "\n",
    "    fig.xaxis.ticker = window_sizes\n",
    "    fig.x_range = bkmod.Range1d(100, 10000)\n",
    "    if title is None:\n",
    "        title = sample_query\n",
    "    fig.title = title\n",
    "    bkplt.show(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e99de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "ag3.plot_h12_calibration(contig=contig, analysis=analysis, sample_query=query_ghana, \n",
    "                    sample_sets=release, cohort_size=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41d8b638",
   "metadata": {},
   "outputs": [],
   "source": [
    "    ds_haps = ag3.haplotypes(\n",
    "        region=contig, \n",
    "        analysis=analysis, \n",
    "        sample_query=sample_query, \n",
    "        sample_sets=sample_sets, \n",
    "        cohort_size=cohort_size, \n",
    "    )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd22dc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def h12_gwss(contig, analysis, sample_query, window_size, \n",
    "             sample_sets, cohort_size=20):\n",
    "\n",
    "    ds_haps = ag3.haplotypes(\n",
    "        region=contig, \n",
    "        analysis=analysis, \n",
    "        sample_query=sample_query, \n",
    "        sample_sets=sample_sets, \n",
    "        cohort_size=cohort_size, \n",
    "    )\n",
    "    \n",
    "    gt = allel.GenotypeDaskArray(ds_haps['call_genotype'].data)\n",
    "    ht = gt.to_haplotypes().compute()\n",
    "    \n",
    "    pos = ds_haps[\"variant_position\"].values\n",
    "\n",
    "    h1, h12, h123, h2_h1 = allel.moving_garud_h(ht, size=window_size)\n",
    "\n",
    "    x = allel.moving_statistic(pos, statistic=np.mean, size=window_size)\n",
    "\n",
    "    return x, h12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8e84a6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "x, h12 = ag3.h12_gwss(contig=contig, analysis=analysis, sample_query=query_ghana, window_size=1000, sample_sets=release)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99bfa4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4671418c",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(h12) == len(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d213761f",
   "metadata": {},
   "outputs": [],
   "source": [
    "h12[11353]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f81dbd33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_h12_gwss(contig, analysis, sample_query, window_size, \n",
    "                  sample_sets, downsample=30, seed=42, \n",
    "                  title=None, width=800, height=200):\n",
    "    \n",
    "    import bokeh.plotting as bkplt\n",
    "    import bokeh.models as bkmod\n",
    "    import bokeh.layouts as bklay\n",
    "\n",
    "    # compute H12\n",
    "    x, h12 = h12_gwss(\n",
    "        contig=contig, \n",
    "        analysis=analysis, \n",
    "        sample_query=sample_query, \n",
    "        sample_sets=sample_sets,\n",
    "        window_size=window_size\n",
    "    )\n",
    "\n",
    "    # determine X axis range\n",
    "    x_min = x[0]\n",
    "    x_max = x[-1]\n",
    "    x_range = bkmod.Range1d(x_min, x_max, bounds=\"auto\")\n",
    "\n",
    "    # create a figure\n",
    "    xwheel_zoom = bkmod.WheelZoomTool(dimensions=\"width\", maintain_focus=False)\n",
    "    fig1 = bkplt.figure(\n",
    "        title=f\"{sample_query}\",\n",
    "        tools=[\"xpan\", \"xzoom_in\", \"xzoom_out\", xwheel_zoom, \"reset\"],\n",
    "        active_scroll=xwheel_zoom,\n",
    "        active_drag=\"xpan\",\n",
    "        plot_width=width,\n",
    "        plot_height=height,\n",
    "        toolbar_location=\"above\",\n",
    "        x_range=x_range,\n",
    "        y_range=(0, 1),\n",
    "    )\n",
    "\n",
    "    # plot H12\n",
    "    fig1.circle(\n",
    "        x=x, \n",
    "        y=h12, \n",
    "        size=3,\n",
    "        line_width=0.5,\n",
    "        line_color=\"black\",\n",
    "        fill_color=None,\n",
    "    )\n",
    "\n",
    "    # tidy up the plot\n",
    "    fig1.yaxis.axis_label = \"H12\"\n",
    "    fig1.yaxis.ticker = [0, 1]\n",
    "    fig1.xaxis.visible = False\n",
    "\n",
    "    # plot genes\n",
    "    fig2 = ag3.plot_genes(\n",
    "        region=contig,\n",
    "        width=width,\n",
    "        height=100,\n",
    "        x_range=fig1.x_range,\n",
    "        show=False,\n",
    "    )\n",
    "\n",
    "    # combine plots into a single figure\n",
    "    fig = bklay.gridplot(\n",
    "        [fig1, fig2], ncols=1, toolbar_location=\"above\", merge_tools=True\n",
    "    )\n",
    "\n",
    "    bkplt.show(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b76b277",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "ag3.plot_h12_gwss(contig=contig, analysis=analysis, sample_query=query_ghana, window_size=1000, sample_sets=release, title = 'Chips')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08e2f44a",
   "metadata": {},
   "outputs": [],
   "source": [
    "check = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ab4f5fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = str(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11eec779",
   "metadata": {},
   "outputs": [],
   "source": [
    "check[a] = 'Tree'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4f10433",
   "metadata": {},
   "outputs": [],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "293c669e",
   "metadata": {},
   "outputs": [],
   "source": [
    "check[100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988004d2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
